packages/public/server/src/module/Alias/src/Controller/AliasController.php

```diff
@@ -22,71 +22,149 @@
  */
 namespace Alias\Controller;
 
-use Alias\Exception\AliasNotFoundException;
-use Alias\Exception\CanonicalUrlNotFoundException;
 use Alias;
+use Alias\AliasManagerInterface;
+use Alias\Controller\Plugin\Url;
+use Instance\Manager\InstanceManagerInterface;
+use Normalizer\NormalizerInterface;
+use Uuid\Exception\NotFoundException;
+use Uuid\Manager\UuidManagerInterface;
 use Zend\Http\Request;
 use Zend\Mvc\Controller\AbstractActionController;
-use Zend\Stdlib\ArrayUtils;
+use Zend\Mvc\Router\RouteInterface;
+use Zend\View\Model\JsonModel;
+use Zend\View\Model\ViewModel;
 
 class AliasController extends AbstractActionController
 {
-    use Alias\AliasManagerAwareTrait,
-        \Instance\Manager\InstanceManagerAwareTrait;
+    /** @var AliasManagerInterface */
+    private $aliasManager;
+    /** @var InstanceManagerInterface */
+    private $instanceManager;
+    /** @var UuidManagerInterface */
+    private $uuidManager;
+    /** @var NormalizerInterface */
+    private $normalizer;
+    /** @var RouteInterface */
+    private $router;
 
-    /**
-     * @var mixed
-     */
-    protected $router;
+    public function __construct(
+        AliasManagerInterface $aliasManager,
+        InstanceManagerInterface $instanceManager,
+        UuidManagerInterface $uuidManager,
+        NormalizerInterface $normalizer,
+        RouteInterface $router
+    ) {
+        $this->aliasManager = $aliasManager;
+        $this->instanceManager = $instanceManager;
+        $this->uuidManager = $uuidManager;
+        $this->normalizer = $normalizer;
+        $this->router = $router;
+    }
 
-    public function forwardAction()
+    public function resolveAction()
     {
         $alias = $this->params('alias');
-        $instance = $this->getInstanceManager()->getInstanceFromRequest();
 
-        try {
-            $location = $this->aliasManager->findCanonicalAlias(
-                $alias,
-                $instance
-            );
-            $this->redirect()->toUrl($location);
-            $this->getResponse()->setStatusCode(301);
-            return false;
-        } catch (CanonicalUrlNotFoundException $e) {
+        if (preg_match('/^(?<id>\d+)$/', $alias, $matches)) {
+            return $this->resolveUuid($matches['id']);
+        } else {
+            $url = $this->resolveAlias($alias);
+            return $url === null
+                ? $this->notFoundResponse()
+                : $this->routerResponse($url);
         }
+    }
 
+    private function resolveAlias($alias)
+    {
+        if (preg_match('/(?<id>\d+)\//', $alias, $matches)) {
+            try {
+                $object = $this->uuidManager->getUuid($matches['id'], true);
+                $normalized = $this->normalizer->normalize($object);
+                return $this->getUrlOfNormalizedObject($normalized);
```

Comment:
```suggestion
                return $this->resolveUuid($matches['id']);
```

Why don't we use `resolveUuid()` here? To have a fallback "UUID not found, fall through to check if this is a legacy alias"?

Reply:
Yep. `resolveUuid` is only used for `/:id` paths (there we have to resolve via ID and don't want to look up an alias). `resolveAlias` is used for the remaining paths. Also note that the behavior for `/:id` vs. `/:some-alias` is different because `/:id` has the custom logic for injections (which didn't really work when using aliases. Since this isn't a supported use case anyways, I'd leave it like that for now).